### ISSUE TITLE
[Optimization] Avoid $(T\times T)$ masks at large $T$ (Sliding-Window SDPA / FA2 path)

### ISSUE THREAD

**Issue by z-a-f** [maida-ai/Research#43]
### Summary
When `W < T` you currently build a full `(1,1,T,T)` bool mask. At large `T` this is heavy. Add a threshold (e.g., `max_mask_T=4096`): below threshold use the cached mask (what you have), above threshold compute attention in **sliding blocks** (no global mask), or optionally call a **FlashAttention-2/3 windowed** path if available later.

### Motivation
Prevents large temporary allocations and unlocks longer contexts on the same GPU.

### Scope

* Add `self.max_mask_T: int = 4096` (configurable).
* In `_compute_attention`:

  * If `W ≥ T`: keep current causal fast-path.
  * Else if `T ≤ max_mask_T`: keep current cached mask path.
  * Else: **blockwise** SDPA: iterate over target blocks and attend to their left window only (no global mask).

### Implementation Sketch (blockwise SDPA)

```python
def _compute_attention(self, x: torch.Tensor, _unused=None):
    B, T, d = x.shape; d_h = d // self.n_heads
    Q = self.W_Q(x).view(B,T,self.n_heads,d_h).transpose(1,2)
    K = self.W_K(x).view(B,T,self.n_heads,d_h).transpose(1,2)
    V = self.W_V(x).view(B,T,self.n_heads,d_h).transpose(1,2)
    p = self.drop.p if self.training else 0.0

    if self.window_size >= T:
        out = F.scaled_dot_product_attention(Q,K,V, attn_mask=None, dropout_p=p, is_causal=True)
    elif T <= self.max_mask_T:
        mask = self._get_window_mask(T, self.window_size, x.device)
        out = F.scaled_dot_product_attention(Q,K,V, attn_mask=mask, dropout_p=p, is_causal=False)
    else:
        # sliding blocks over target time dimension
        W = self.window_size
        out_chunks = []
        # choose a block size (e.g., BS = W or 2W)
        BS = min(2*W, T)
        for t0 in range(0, T, BS):
            t1 = min(T, t0 + BS)
            # Keys/values span [max(0,t1-W), t1)
            ks = max(0, t1 - W)
            Qb = Q[:, :, t0:t1, :]                  # (B,h,BS,dh)
            Kb = K[:, :, ks:t1, :]                  # (B,h,KB,dh)
            Vb = V[:, :, ks:t1, :]
            # Build a small causal mask inside the block:
            # target positions are t0..t1-1, allowed keys <= each position
            # local indices:
            Tb = t1 - t0; Kb_len = t1 - ks
            i = torch.arange(Tb, device=x.device)[:, None]
            j = torch.arange(Kb_len, device=x.device)[None, :]
            # position of key j corresponds to global time ks+j
            # we need j_global <= i_global -> (ks+j) <= (t0+i)  => j <= i + (t0-ks)
            shift = (t0 - ks)
            band = j <= (i + shift)
            mask_b = band.view(1,1,Tb,Kb_len)
            out_b = F.scaled_dot_product_attention(Qb,Kb,Vb, attn_mask=mask_b, dropout_p=p, is_causal=False)
            out_chunks.append(out_b)
        out = torch.cat(out_chunks, dim=2)

    out = out.transpose(1,2).contiguous().view(B,T,d)
    return self.drop(self.W_O(out))
```

### Tests

* T=512, W=256 uses cached-mask path; T=16384, W=256 uses blockwise path (no OOM).
* Correctness: change a token beyond `W` on the left doesn’t affect the current position.
* Compare outputs of cached-mask vs. blockwise for small T (should be close).

### Acceptance Criteria

* [ ] No $(T\times T)$ allocations when `T > max_mask_T`.
* [ ] Equal or better throughput at large T.
* [ ] Unit tests cover both paths.

### Rollout notes

* Keep new behavior **behind flags** where practical (`ssm_impl`, `tile_size`, `num_kv_heads`, `proj_stride`, `overlap_streams`, `max_mask_T`, `compile`).
* Add quick **unit tests** (shape, state continuity, streaming equivalence).
* In each PR description, include a **tiny benchmark snippet** (tokens/sec @ T=512 or 1024) to document gains.

### Context

Even a bool mask blows up VRAM (e.g., $16k\times 16k \approx 256 \text{MB}$). Add a threshold (e.g., `max_mask_T=4096`)-above it, do **sliding block SDPA** (compute attention in overlapping blocks of size $W$, no global mask) or call **FlashAttention-2/3** with its native `window_size=(W_left,W_right)` which avoids any $(T\times T)$ tensor. (https://github.com/Dao-AILab/flash-attention)

```python
# in _compute_attention
if self.window_size >= T:
    attn_mask, is_causal = None, True
elif T <= self.max_mask_T:
    attn_mask, is_causal = self._get_window_mask(T, self.window_size, x.device), False
else:
    # option A: blockwise SDPA over sliding windows (no full mask)
    # option B: if FA2 available, call flash_attn with window_size=(W,0)
```

Also keep the **SDPA rule**: never set `attn_mask` and `is_causal=True` simultaneously; it de-optimizes and can error. (Your code already does this correctly.) (https://docs.pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html)
